src/gluonts/transform/feature.py

```diff
@@ -433,3 +433,198 @@ def map_transform(self, data: DataEntry, is_train: bool) -> DataEntry:
+    rolling_agg:
+        Boolean indicating if the aggregation should be done in a centered rolling
+        window fashion (default) or by calendar dates.
+    """
+
+    @validated()
+    def __init__(
+        self,
+        target_field: str,
+        output_field: str,
+        pred_length: int,
+        base_freq: str,
+        agg_freq: str,
+        agg_lags: List[int],
+        agg_fun: str = "mean",
+        rolling_agg: bool = True,
+        dtype: DType = np.float32,
+    ) -> None:
+        self.pred_length = pred_length
+        self.target_field = target_field
+        self.feature_name = output_field
+        self.base_freq = base_freq
+        self.agg_freq = agg_freq
+        self.agg_lags = agg_lags
+        self.agg_fun = agg_fun
+        self.rolling_agg = rolling_agg
+        self.dtype = dtype
+
+        self.ratio = pd.Timedelta(self.agg_freq) / pd.Timedelta(self.base_freq)
+        assert (
+            self.ratio.is_integer() and self.ratio >= 1
+        ), "The aggregate frequency should be a multiple of the base frequency."
+        self.ratio = int(self.ratio)
+
+        if rolling_agg:
+            self.half_window = (self.ratio - 1) // 2
+            self.valid_lags = [
+                x
+                for x in self.agg_lags
+                if x > (self.pred_length - 1 + self.half_window) / self.ratio
+            ]
+        else:
+            self.valid_lags = [
+                x
+                for x in self.agg_lags
+                if x > np.ceil((self.pred_length - 1) / self.ratio)
+            ]
+        if set(self.agg_lags) - set(self.valid_lags):
+            print(
+                f"The aggregate lags {set(self.agg_lags) - set(self.valid_lags)} "
+                f"of frequency {self.agg_freq} are ignored."
+            )
+
+    def map_transform(self, data: DataEntry, is_train: bool) -> DataEntry:
+        assert self.base_freq == data["start"].freq
+
+        # convert to pandas Series for easier indexing and aggregation
+        if is_train:
+            pd_ts = pd.Series(
+                data[self.target_field],
+                index=pd.date_range(
+                    data["start"],
+                    periods=len(data[self.target_field]),
+                    freq=self.base_freq,
+                ),
+            )
+        else:
+            pd_ts = pd.Series(
+                np.concatenate(
+                    [
+                        data[self.target_field],
+                        np.zeros(shape=(self.pred_length,)),
+                    ],
+                    axis=0,
+                ),
+                index=pd.date_range(
+                    data["start"],
+                    periods=len(data[self.target_field]) + self.pred_length,
+                    freq=self.base_freq,
+                ),
+            )
+
+        if not self.rolling_agg:
+            # compute how many time stamps are in the last (potentially not full) aggregation window
```

Comment:
I don't fully understand the details for the non-rolling case, but I wonder if it could be simplified.
Perhaps we can use some pandas logic such as

```python
pd_ts.resample('1H').mean().repeat(60).values
```

to create an array of lagged values that behave in this non-rolling manner. Maybe this is missing some edge cases though, so feel free to ignore if this is missing something.

Reply:
About `resample` I do not directly apply it because it aggregates windows that are not fully observed. For example if a time series starts at 12.50, with `1H` resample we will get a value for 12 that is the aggregate of the last 10min and this can be way off. Therefore, I cut the first and last windows IF they are not full (last window does not matter eitherway value-wise because you cannot use lag 0). Then I apply the `resample` method to the fully observed windows (see line 595).

About `repeat`, this was an option but in the end I decided not to do it. One thing that we cannot avoid is the indexes that point to the correct lags of each timestep which should be K x T (number of lags, length of time series). These indexes point directly to the correct place of the aggregated time series without repeating the timesteps. This is just to reduce the computational and memory burdens of the transformation. Having said that, by repeating the two cases (rolling and not rolling) could potentially be a bit more unified and the code a bit more clean (with the tradeoff of repeating the aggregate time series). I am happy to do this change if you think it is better.